File: Websites/Glory2Him.WebApp.Tests.Acceptance/Models/ApprovalSettings/ApprovalSetting.cs
// ────────────────────────────────────────────────────────────────────────────────
// Copyright (c) Glory 2 Him. All rights reserved.
// Licensed under the Glory 2 Him Software License (G2HSL).
// See License.txt in the project root for full license information.
// FREE TO USE TO HELP SHARE THE GOSPEL
// John 14:6 (NIV) "Jesus answered, ‘I am the way and the truth and the life.
//                  No one comes to the Father except through me.’"
// https://john.bible/john-14-6
// If Jesus is who He said He is, what does that mean for you, today?
// ────────────────────────────────────────────────────────────────────────────────

using System;
using Glory2Him.Core.Models.Enums;

namespace Glory2Him.WebApp.Tests.Acceptance.Models.ApprovalSettings
{
    /// <summary>
    /// The wire shape of an approval policy row.
    ///
    /// <para><b>No approval fields</b>, unlike every other model in this project. §7.5 entry 9
    /// makes <c>ApprovalSetting</c> approvable only <i>"if policy changes require approval"</i>,
    /// a conditional never taken up — the entity carries no <c>ApprovalStatus</c>,
    /// <c>IsPublished</c>, <c>PublishDate</c> or bypass pair, so a model copied from a content
    /// entity would declare five properties the API never sends.</para>
    /// </summary>
    public class ApprovalSetting
    {
        public Guid Id { get; set; }

        // The scope. EntityType alone is the per-type DEFAULT; EntityType plus ContentType is an
        // override. Each has its own filtered unique index, so exactly one row can occupy either
        // scope (§8.4 policy resolution depends on that).
        public EntityType EntityType { get; set; }
        public ContentType? ContentType { get; set; }

        // The policy itself (§8.2).
        public bool RequireApprovals { get; set; }
        public int RequiredNumberOfApprovals { get; set; }
        public bool AutoApproveIfAllApprovalRequirementsMet { get; set; }
        public bool AllowSelfApproval { get; set; }
        public bool BlockOnReject { get; set; }
        public bool BlockOnZeroApprovalScore { get; set; }
        public bool RequireReapprovalOnChange { get; set; }
        public bool RequireReviewCommentResolutionBeforeApprovals { get; set; }
        public bool DoNotAllowBypassingSettings { get; set; }

        public string CreatedBy { get; set; }
        public DateTimeOffset CreatedWhen { get; set; }
        public string UpdatedBy { get; set; }
        public DateTimeOffset UpdatedWhen { get; set; }
        public string DeletedBy { get; set; }
        public DateTimeOffset? DeletedWhen { get; set; }
        public bool IsDeleted { get; set; }
        public string DeletionReason { get; set; }
    }
}

File: Websites/Glory2Him.WebApp.React/src/models/foundations/approvalSettings/approvalSetting.ts
import { ContentType } from '../contentItemSettings/contentType';

// Mirrors Glory2Him.Core.Models.Enums.EntityType, including its numbering: the wire carries the
// NUMBER (the host registers no JsonStringEnumConverter), and an approval setting's scope is
// keyed on it, so these values are a contract rather than a convenience.
export enum EntityType {
    ContentItem = 0,
    Tag = 1,
    Reaction = 2,
    BibleReference = 3,
    Comment = 4,
    Link = 5,
    Attachment = 6,
    Association = 7,
}

// Declaration order, for the pickers. Object.keys over a numeric enum yields the reverse mapping
// alongside the members, so the list is stated rather than derived.
export const entityTypeMembers: ReadonlyArray<EntityType> = [
    EntityType.ContentItem,
    EntityType.Tag,
    EntityType.Reaction,
    EntityType.BibleReference,
    EntityType.Comment,
    EntityType.Link,
    EntityType.Attachment,
    EntityType.Association,
];

export const entityTypeLabels: Readonly<Record<EntityType, string>> = {
    [EntityType.ContentItem]: 'Content item',
    [EntityType.Tag]: 'Tag',
    [EntityType.Reaction]: 'Reaction',
    [EntityType.BibleReference]: 'Bible reference',
    [EntityType.Comment]: 'Comment',
    [EntityType.Link]: 'Link',
    [EntityType.Attachment]: 'Attachment',
    [EntityType.Association]: 'Association',
};

// ONLY ContentItem CARRIES A CONTENT TYPE (design §8.4). Every other entity type must leave it
// null, and the database enforces that with a CHECK constraint rather than the service — so a
// bad pair comes back as a dependency failure with no field to hang it on. The form prevents it
// instead of explaining it afterwards.
export const allowsContentTypeScope = (entityType: EntityType): boolean =>
    entityType === EntityType.ContentItem;

// Wire shape of api/ApprovalSettings, camelCased by the host's default System.Text.Json policy.
//
// THE WHOLE ROW GOES BACK ON A SAVE. PUT binds an ApprovalSetting and validates Id, CreatedBy and
// CreatedWhen against storage BEFORE reading it, so an edit is the fetched row with the policy
// fields changed — never a fresh object. UpdatedBy and UpdatedWhen are server-stamped and
// whatever a caller sends for them is ignored.
export type ApprovalSetting = {
    // Supplied by the CALLER on create: the service never generates one, and refuses an empty
    // Guid. There is no POST that mints an id for you.
    id: string;

    // ── Scope (§8.4) ──────────────────────────────────────────────────────────
    entityType: EntityType;

    // Null means "every content type of this entity type" — the entity-type default tier. A row
    // naming a content type beats it for that type.
    contentType: ContentType | null;

    // ── Policy ────────────────────────────────────────────────────────────────
    requireApprovals: boolean;
    requiredNumberOfApprovals: number;
    autoApproveIfAllApprovalRequirementsMet: boolean;
    allowSelfApproval: boolean;
    blockOnReject: boolean;
    blockOnZeroApprovalScore: boolean;
    requireReapprovalOnChange: boolean;
    requireReviewCommentResolutionBeforeApprovals: boolean;
    doNotAllowBypassingSettings: boolean;

    // ── Audit ─────────────────────────────────────────────────────────────────
    // Carried rather than displayed only: the save round-trips CreatedBy and CreatedWhen, which
    // the foundation compares against storage before it will accept the write.
    createdBy: string;
    createdWhen: string;
    updatedBy: string;
    updatedWhen: string;
    isDeleted: boolean;
};

// What a NEW row opens on: the HOUSE POLICY, the same nine values ApprovalSettingSeedData writes
// for every entity-type default. A content-type row an administrator adds narrows a seeded
// default, so it opens matching that default and the administrator changes only what they
// mean to — a form that opened looser than the row it overrides would make the ninth policy
// quietly weaker than the eight.
export const newApprovalSetting = (id: string): ApprovalSetting => ({
    id,
    entityType: EntityType.ContentItem,
    contentType: null,
    requireApprovals: true,
    requiredNumberOfApprovals: 2,
    autoApproveIfAllApprovalRequirementsMet: false,
    allowSelfApproval: false,
    blockOnReject: true,
    blockOnZeroApprovalScore: true,
    requireReapprovalOnChange: true,
    requireReviewCommentResolutionBeforeApprovals: true,
    doNotAllowBypassingSettings: false,
    createdBy: '',
    createdWhen: '',
    updatedBy: '',
    updatedWhen: '',
    isDeleted: false
});
